ostree/ostree_transport.go

```diff
@@ -0,0 +1,177 @@
+package ostree
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/containers/image/docker/policyconfiguration"
+	"github.com/containers/image/docker/reference"
+	"github.com/containers/image/types"
+)
+
+// Transport is an ImageTransport for ostree paths.
+var Transport = ostreeTransport{}
+
+type ostreeTransport struct{}
+
+func (t ostreeTransport) Name() string {
+	return "ostree"
+}
+
+// ValidatePolicyConfigurationScope checks that scope is a valid name for a signature.PolicyTransportScopes keys
+// (i.e. a valid PolicyConfigurationIdentity() or PolicyConfigurationNamespaces() return value).
+// It is acceptable to allow an invalid value which will never be matched, it can "only" cause user confusion.
+// scope passed to this function will not be "", that value is always allowed.
+func (t ostreeTransport) ValidatePolicyConfigurationScope(scope string) error {
+	if strings.Index(scope, "@") < 0 {
+		return errors.Errorf("Invalid scope %s: Must include a repo", scope)
+	}
+	return nil
+}
+
+// ostreeReference is an ImageReference for ostree paths.
+type ostreeReference struct {
+	ref   string
+	repo  string
+	image reference.Named
+}
+
+func (t ostreeTransport) ParseReference(ref string) (types.ImageReference, error) {
+	var repo = ""
+	var imageStr = ""
+	s := strings.SplitN(ref, "@", 2)
+	if len(s) == 1 {
+		imageStr, repo = s[0], "/ostree/repo"
+	} else {
+		imageStr, repo = s[0], s[1]
+	}
+
+	var imageTagged reference.Named
+	s = strings.SplitN(imageStr, ":", 2)
+
+	image, err := reference.WithName(s[0])
```

Comment:
The point is, what is the _semantics_ of hostname-less `busybox`?

That notably affects whether `docker/policyconfiguration` is relevant, and whether `DockerReference` should return the result of parsing, because these interfaces _do_ expect a `reference.Named` with the docker/distribution upstream semantics, i.e. that every reference implicitly contains a hostname, that `docker.io/[library/]` may be implicit / canonicalized, and that `.Name()` and `.String()` return canonical forms which can be compared for equality.

Per the docker/distribution upstream semantics, `busybox` and `docker.io/library/busybox:latest` are two ways to refer to the same image, and `policyConfiguration.DockerReferenceNamespaces(reference.ParseNormalizedNamed(input))` correspondingly returns the same sequence of namespaces (namely [`docker.io/library/busybox`, `docker.io/library`, `docker.io`]). If the ostree image names have a different semantics, such that `busybox` has no logical parent namespace, then the `PolicyConfiguration*` methods should be adjusted accordingly, perhaps avoiding `docker/reference` entirely and maybe doing its own parsing using… `strings.Split`, or its own regexps/rules, I guess.

Also, `.DockerReference()` is used to figure out how to sign images; if that returns a `"busybox"` reference for an ostree image, this value is used to create a signature, consumers of signatures will parse it using `reference.ParseNormalizedNamd` and consider such a signature acceptable when the consumer wants a `docker.io/library/busybox` image. If the semantics of the `busybox` value in ostree is in fact different, then `.DockerReference()` must return the right value, or perhaps `nil` if the image identity can’t be expressed using the upstream docker/distribution `refererence.Named` semantics.